### ISSUE TITLE
Livelock issue in MpscLinkedQueue.java

### ISSUE THREAD

**Issue by shawn11ZX** [netty/netty#5297]
@normanmaurer We still experience the same issue as: Livelock issue in MpscLinkedQueue.java #3675
Our version is netty-all-4.1.0.CR7.jar, which already contains the fix : cf66edb3a10fa2bb60802e91f14565b551c08f97 

The call stack is:

`
"GameLoginScheduledThread" prio=10 tid=0x00007f01d15f1800 nid=0xa2ed runnable [0x00007effdf4f3000]
   java.lang.Thread.State: RUNNABLE
        **at io.netty.util.internal.MpscLinkedQueue.peekNode(MpscLinkedQueue.java:106)
        at io.netty.util.internal.MpscLinkedQueue.size(MpscLinkedQueue.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.pendingTasks(SingleThreadEventExecutor.java:269)**
        at common.net.NettyMonitor.saveAndReset(NettyMonitor.java:27)
        at common.statistics.StatisticManager.reportAndReset(StatisticManager.java:67)
        at gamelogin.Manager$2.run(Manager.java:211)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
`

**Comment by normanmaurer** [comment]
@shawn11ZX are you sure that you really call size() from the same EventLoop as it is assigned too ?

**Comment by normanmaurer** [comment]
That said I think I can make it safe from any thread....

**Comment by Jire** [comment]
I am having the same issue occur at random.

**Comment by shawn11ZX** [comment]
No, I call pendingTasks from a third thread which is used to monitor netty status. The code is as following: 
```
public class NettyEventLoopGroupMonitor implements StatisticItem {
	EventLoopGroup group;
	public NettyEventLoopGroupMonitor(EventLoopGroup group)
	{
		this.group = group;
	}

	@Override
	public void saveAndReset(String name, Map<String, Object> output) {
		Set<EventExecutor> executors = group.children();
		int tasks = 0;
		for (EventExecutor executor: executors)
		{
			if (executor instanceof SingleThreadEventExecutor) {
				SingleThreadEventExecutor e = (SingleThreadEventExecutor) executor;
				tasks += e.pendingTasks();
				
			}
			
		}
		
		output.put("netty_" + name, tasks);
	}
}
```

**Comment by normanmaurer** [comment]
Fixed via: https://github.com/netty/netty/pull/5309

**Comment by normanmaurer** [comment]
fixed by https://github.com/netty/netty/pull/5309

**Comment by shayshim** [comment]
Norman, what if thread of SingleThreadEventExecutor is unexpectedly blocked/long calculates somewhere? So all producers might continue to aggressively fill its queue, which ultimately can cause OOME, but the thread is not available to answer the queue size. So even different thread which called Nio/EpollEventLoop.pendingTasks will have to syncUninterruptibly without being able to report the size of the queue, possibly for long time. When the consumer thread answers the size - it might already be after it consumed most of the queue, so we missed the size at the time we asked for it.
I know size doesn't has to be accurate for such a queue, but I think it is important for analysis purpose that the thread which asks the size will be independent of the thread which consumes the queue. We need to know the size especially if the consumer is blocked somewhere and producers continue to fill its queue. Don't want to wait and depend on the consumer in such case.

Except the OOME, similar behavior to what I described actually happened to us, while we tested new statistic for this queue.

Did you consider using LongAdder for maintaining the size? It maybe could be used inside the queue.
